synapse/handlers/room_member.py

```diff
@@ -710,9 +777,96 @@ def _lookup_3pid(self, id_server, medium, address):
                 return data["mxid"]
 
         except IOError as e:
-            logger.warn("Error from identity server lookup: %s" % (e,))
+            logger.warning("Error from v1 identity server lookup: %s" % (e,))
+
+        return None
+
+    @defer.inlineCallbacks
+    def _lookup_3pid_v2(
+        self, id_server, id_access_token, medium, address, hash_details
+    ):
+        """Looks up a 3pid in the passed identity server using v2 lookup.
+
+        Args:
+            id_server (str): The server name (including port, if required)
+                of the identity server to use.
+            id_access_token (str): The access token to authenticate to the identity server with
+            medium (str): The type of the third party identifier (e.g. "email").
+            address (str): The third party identifier (e.g. "foo@example.com").
+            hash_details (dict[str, str|list]): A dictionary containing hashing information
+                provided by an identity server.
+
+        Returns:
+            Deferred[str|None]: the matrix ID of the 3pid, or None if it is not recognised.
+        """
+        # Extract information from hash_details
+        supported_lookup_algorithms = hash_details.get("algorithms")
+        lookup_pepper = hash_details.get("lookup_pepper")
+        if (
+            not supported_lookup_algorithms
+            or not isinstance(supported_lookup_algorithms, list)
+            or not lookup_pepper
+            or not isinstance(lookup_pepper, str)
+        ):
+            raise SynapseError(
```

Comment:
there's no value to raising a SynapseError with a 500 error code: it may as well be a regular `Exception`. I'm a bit dubious that this is the correct behaviour though: it's not synapse's fault that the id server is giving incorrect values, so this should be `None` or 421 or 4xx or something

Reply:
Changed to 421. I think it's best to tell the client when they're using an identity server that's giving them malformed results, such that they can know to change it (since they can do that these days).

Reply:
Changed to 400 as per privacy-dev discussion.